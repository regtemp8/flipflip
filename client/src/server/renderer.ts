import createCache, { type EmotionCache } from '@emotion/cache'
import { configureStore } from '@reduxjs/toolkit'
import { AppStorage, SystemConstants } from 'flipflip-common'
import createAppReducer from '../store/app/slice'
import createAudioReducer from '../store/audio/slice'
import createCaptionScriptReducer from '../store/captionScript/slice'
import createClipReducer from '../store/clip/slice'
import createLibrarySourceReducer from '../store/librarySource/slice'
import createPlaylistReducer from '../store/playlist/slice'
import createSceneReducer from '../store/scene/slice'
import createSceneGroupReducer from '../store/sceneGroup/slice'
import createTagReducer from '../store/tag/slice'
import playersSlice from '../store/player/slice'
import videoClipperSlice from '../store/videoClipper/slice'
import scenePickerSlice from '../store/scenePicker/slice'
import captionScriptorSlice from '../store/captionScriptor/slice'
import sceneDetailSlice from '../store/sceneDetail/slice'
import audioLibrarySlice from '../store/audioLibrary/slice'
import sourceScraperSlice from '../store/sourceScraper/slice'
import createConstantsReducer from '../store/constants/slice'
import { fromAppStorage } from '../store/app/convert'
import createDisplayReducer from '../store/display/slice'
import createDisplayViewReducer from '../store/displayView/slice'
import createScenePlaylistItemReducer from '../store/scenePlaylistItem/slice'
import createDisplayPlaylistItemReducer from '../store/displayPlaylistItem/slice'

export function createEmotionCache(): EmotionCache {
  return createCache({ key: 'css' })
}

export function createReduxStore(
  appStorage?: AppStorage,
  constants?: SystemConstants
) {
  const state = appStorage != null ? fromAppStorage(appStorage) : undefined
  return configureStore({
    reducer: {
      app: createAppReducer(state?.app),
      audio: createAudioReducer(state?.audio),
      captionScript: createCaptionScriptReducer(state?.captionScript),
      clip: createClipReducer(state?.clip),
      librarySource: createLibrarySourceReducer(state?.librarySource),
      playlist: createPlaylistReducer(state?.playlist),
      displayPlaylistItem: createDisplayPlaylistItemReducer(
        state?.displayPlaylistItem
      ),
      scenePlaylistItem: createScenePlaylistItemReducer(
        state?.scenePlaylistItem
      ),
      scene: createSceneReducer(state?.scene),
      sceneGroup: createSceneGroupReducer(state?.sceneGroup),
      display: createDisplayReducer(state?.display),
      displayView: createDisplayViewReducer(state?.displayView),
      tag: createTagReducer(state?.tag),
      constants: createConstantsReducer(constants),
      // components
      players: playersSlice,
      videoClipper: videoClipperSlice,
      scenePicker: scenePickerSlice,
      captionScriptor: captionScriptorSlice,
      sceneDetail: sceneDetailSlice,
      audioLibrary: audioLibrarySlice,
      sourceScraper: sourceScraperSlice
    }
  })
}
