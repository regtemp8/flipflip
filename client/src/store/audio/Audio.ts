import { copy, TF } from 'flipflip-common'

export default interface Audio {
  [key: string]: number | number[] | string | boolean | undefined

  id: number
  url?: string
  marked: boolean
  tags: number[] // Array of Tag IDs
  volume: number
  speed: number
  stopAtEnd: boolean
  nextSceneAtEnd: boolean
  tick: boolean
  tickMode: string
  tickDelay: number
  tickMinDelay: number
  tickMaxDelay: number
  tickSinRate: number
  tickBPMMulti: number
  bpm: number
  thumb?: string
  name?: string
  artist?: string
  album?: string
  trackNum?: number
  duration?: number
  comment?: string
  playedCount: number
}

export const initialAudio: Audio = {
  id: 0,
  marked: false,
  tags: [],
  volume: 100,
  speed: 10,
  stopAtEnd: false,
  nextSceneAtEnd: false,
  tick: false,
  tickMode: TF.constant,
  tickDelay: 1000,
  tickMinDelay: 500,
  tickMaxDelay: 5000,
  tickSinRate: 100,
  tickBPMMulti: 10,
  bpm: 0,
  playedCount: 0
}

export function newAudio(init?: Partial<Audio>) {
  const audio = Object.assign(copy<Audio>(initialAudio), init)
  if (audio.tickMode === 'at.constant') {
    audio.tickMode = 'tf.c'
  } else if (audio.tickMode === 'at.random') {
    audio.tickMode = 'tf.random'
  } else if (audio.tickMode === 'at.sin') {
    audio.tickMode = 'tf.sin'
  } else if (audio.tickMode === 'at.scene') {
    audio.tickMode = 'tf.scene'
  }

  return audio
}
