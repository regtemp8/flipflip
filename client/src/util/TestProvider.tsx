import * as React from 'react'
import { Provider } from 'react-redux'
import { createTheme } from '@mui/material'
import { StyledEngineProvider, ThemeProvider } from '@mui/material/styles'
import defaultTheme from 'flipflip-common'

export default class TestProvider extends React.Component {
  readonly props: {
    children: any
    store: any
  }

  disableTransitions = {
    defaultProps: {
      disablePortal: true,
      hideBackdrop: true,
      TransitionComponent: ({ children }) => children
    }
  }

  render() {
    const theme = createTheme({
      ...defaultTheme,
      // disable Transitions when runing tests, throws errors otherwise.
      components: {
        MuiMenu: this.disableTransitions,
        MuiPopover: this.disableTransitions,
        MuiDialog: this.disableTransitions,
        MuiModal: this.disableTransitions
      }
    } as any)
    return (
      <Provider store={this.props.store}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>{this.props.children}</ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    )
  }
}
