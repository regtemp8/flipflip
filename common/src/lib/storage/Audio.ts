import { TF } from '../const';
import { copy } from '../utils';

import { type Tag } from './Tag';

export type Audio = {
  id: number;
  url?: string;
  marked: boolean;
  tags: Tag[];
  volume: number;
  speed: number;
  stopAtEnd: boolean;
  nextSceneAtEnd: boolean;
  tick: boolean;
  tickMode: string;
  tickDelay: number;
  tickMinDelay: number;
  tickMaxDelay: number;
  tickSinRate: number;
  tickBPMMulti: number;
  bpm: number;
  thumb?: string;
  name?: string;
  artist?: string;
  album?: string;
  trackNum?: number;
  duration?: number;
  comment?: string;
  playedCount: number;
};

export const initialAudio: Audio = {
  id: 0,
  marked: false,
  tags: [],
  volume: 100,
  speed: 10,
  stopAtEnd: false,
  nextSceneAtEnd: false,
  tick: false,
  tickMode: TF.constant,
  tickDelay: 1000,
  tickMinDelay: 500,
  tickMaxDelay: 5000,
  tickSinRate: 100,
  tickBPMMulti: 10,
  bpm: 0,
  playedCount: 0,
};

export function newAudio(init?: Partial<Audio>): Audio {
  const tickModes: Record<string, string> = {};
  tickModes['at.constant'] = 'tf.c';
  tickModes['at.random'] = 'tf.random';
  tickModes['at.sin'] = 'tf.sin';
  tickModes['at.scene'] = 'tf.scene';

  const audio = Object.assign(copy<Audio>(initialAudio), init);
  const tickMode = tickModes[audio.tickMode];
  if (tickMode != null) {
    audio.tickMode = tickMode;
  }

  return audio;
}
