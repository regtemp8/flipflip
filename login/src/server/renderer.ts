import createCache, { type EmotionCache } from "@emotion/cache";

export function createEmotionCache(nonce?: string): EmotionCache {
  return createCache({ key: "css", nonce });
}
