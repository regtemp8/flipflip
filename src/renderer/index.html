<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>FlipFlip</title>
    <!-- Responsive meta tag -- ensures proper rendering and touch zooming for all devices -->
    <meta name="viewport" content="minimum-scale=1, initial-scale=1, width=device-width, shrink-to-fit=no" />
    <!-- Roboto Font -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap" />
    <!-- Font Icons -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons" />

    <meta
      http-equiv="Content-Security-Policy"
      content="frame-src https://www.imagefap.com http://www.imagefap.com https://hypno.nimja.com/ http://hypno.nimja.com/; font-src 'self' https://fonts.gstatic.com; style-src 'self' https://fonts.googleapis.com 'unsafe-inline'"
    />
  </head>

  <body>
    <div id="app" class="app"></div>
  </body>
</html>
