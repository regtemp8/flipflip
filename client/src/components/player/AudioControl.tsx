import React, { useEffect, useState, useRef } from 'react'
import Sound from 'react-sound'
import { cx } from '@emotion/css'

import {
  Collapse,
  Grid,
  IconButton,
  Slider,
  type Theme,
  Tooltip,
  Typography
} from '@mui/material'

import { makeStyles } from 'tss-react/mui'

import Forward10Icon from '@mui/icons-material/Forward10'
import Forward5Icon from '@mui/icons-material/Forward5'
import Replay10Icon from '@mui/icons-material/Replay10'
import Replay5Icon from '@mui/icons-material/Replay5'
import PauseIcon from '@mui/icons-material/Pause'
import PlayArrowIcon from '@mui/icons-material/PlayArrow'
import SkipNextIcon from '@mui/icons-material/SkipNext'
import SkipPreviousIcon from '@mui/icons-material/SkipPrevious'
import VolumeDownIcon from '@mui/icons-material/VolumeDown'
import VolumeUpIcon from '@mui/icons-material/VolumeUp'

import { getDuration, getMsRemainder, getTimestamp } from '../../data/utils'
import { RP, TF } from 'flipflip-common'
import SoundTick from './SoundTick'
import { nextScene } from '../../store/scene/thunks'
import { playTrack } from '../../store/audio/slice'
import BaseSlider from '../common/slider/BaseSlider'
import { useAppDispatch, useAppSelector } from '../../store/hooks'
import {
  selectAudioBPM,
  selectAudioUrl,
  selectAudioTick,
  selectAudioTickMode,
  selectAudioTickDelay,
  selectAudioTickMaxDelay,
  selectAudioTickMinDelay,
  selectAudioTickSinRate,
  selectAudioTickBPMMulti,
  selectAudioVolume,
  selectAudioSpeed,
  selectAudioStopAtEnd,
  selectAudioNextSceneAtEnd
} from '../../store/audio/selectors'

const useStyles = makeStyles()((theme: Theme) => ({
  fullWidth: {
    width: '100%'
  },
  noPadding: {
    padding: '0 !important'
  },
  noTransition: {
    transition: 'unset'
  }
}))

export interface AudioControlProps {
  sceneID: number
  audioID: number
  audioEnabled: boolean
  singleTrack: boolean
  lastTrack: boolean
  repeat: string
  scenePaths: any[]
  startPlaying: boolean
  shorterSeek?: boolean
  showMsTimestamp?: boolean
  audioVolumeAction: (value: number) => any
  nextTrack?: () => void
  prevTrack?: () => void
  goBack?: () => void
  onPlaying?: (position: number, duration: number) => void
}

function AudioControl(props: AudioControlProps) {
  const dispatch = useAppDispatch()
  const bpm = useAppSelector(selectAudioBPM(props.audioID))
  const url = useAppSelector(selectAudioUrl(props.audioID))
  const tick = useAppSelector(selectAudioTick(props.audioID))
  const tickMode = useAppSelector(selectAudioTickMode(props.audioID))
  const tickDelay = useAppSelector(selectAudioTickDelay(props.audioID))
  const tickMaxDelay = useAppSelector(selectAudioTickMaxDelay(props.audioID))
  const tickMinDelay = useAppSelector(selectAudioTickMinDelay(props.audioID))
  const tickSinRate = useAppSelector(selectAudioTickSinRate(props.audioID))
  const tickBPMMulti = useAppSelector(selectAudioTickBPMMulti(props.audioID))
  const stopAtEnd = useAppSelector(selectAudioStopAtEnd(props.audioID))
  const nextSceneAtEnd = useAppSelector(
    selectAudioNextSceneAtEnd(props.audioID)
  )
  const volume = useAppSelector(selectAudioVolume(props.audioID))
  const speed = useAppSelector(selectAudioSpeed(props.audioID))

  const [playing, setPlaying] = useState(props.startPlaying)
  const [position, setPosition] = useState(0)
  const [duration, setDuration] = useState(0)
  const [tickState, setTickState] = useState(false)

  const _timeout = useRef<number>()
  const _queueNextTrack = useRef(false)

  useEffect(() => {
    if (playing) {
      tickLoop(true)
    }

    _queueNextTrack.current = false
    return () => {
      if (_timeout.current) {
        clearTimeout(_timeout.current)
      }

      _queueNextTrack.current = false
    }
  }, [])

  useEffect(() => {
    setPosition(0)
    setDuration(0)
  }, [url])

  useEffect(() => {
    if (playing && tick && tickMode !== TF.scene) {
      tickLoop(true)
    } else {
      clearTimeout(_timeout.current)
    }
  }, [playing])

  useEffect(() => {
    if (
      tick &&
      tickMode === TF.scene &&
      props.scenePaths &&
      props.scenePaths.length > 0
    ) {
      if (_queueNextTrack.current) {
        _queueNextTrack.current = false
        props.nextTrack!()
        setPosition(0)
        setDuration(0)
      }

      setTickState(!tickState)
    }
  }, [tick, tickMode, props.scenePaths])

  useEffect(() => {
    if (tick) {
      tickLoop(true)
    }
  }, [tick])

  useEffect(() => {
    if (tick && tickMode === TF.scene) {
      tickLoop(true)
    }
  }, [tickMode])

  const getTimestampFromMs = (ms: number): string => {
    const secs = Math.floor(ms / 1000)
    return getTimestamp(secs)
  }

  const tickLoop = (starting: boolean = false) => {
    if (!starting) {
      if (_queueNextTrack.current) {
        _queueNextTrack.current = false
        props.nextTrack!()
        setPosition(0)
        setDuration(0)
      }

      setTickState(!tickState)
    }
    if (tick) {
      const timeout = getDuration(
        {
          timingFunction: tickMode,
          time: tickDelay,
          timeMax: tickMaxDelay,
          timeMin: tickMinDelay,
          bpmMulti: tickBPMMulti,
          sinRate: tickSinRate
        },
        0,
        bpm
      )
      if (timeout != null) {
        _timeout.current = window.setTimeout(tickLoop, timeout)
        return
      }
    }
    _timeout.current = undefined
  }

  const onChangePosition = (
    e: Event,
    value: number | number[],
    activeThumb: number
  ) => {
    const newPosition: number = Array.isArray(value) ? value[0] : value
    setPosition(newPosition)
  }

  const onFinishedPlaying = () => {
    // Increment play count upon finish
    dispatch(playTrack(props.audioID))

    if (stopAtEnd && props.goBack) {
      props.goBack()
    } else if (nextSceneAtEnd) {
      dispatch(nextScene(props.sceneID))
      setPosition(0)
      setDuration(0)
    } else {
      if (props.repeat === RP.all) {
        if (tick) {
          _queueNextTrack.current = true
        } else {
          if (props.singleTrack) {
            setPosition(1)
          } else {
            props.nextTrack!()
            setPosition(0)
            setDuration(0)
          }
        }
      } else if (props.repeat === RP.one) {
        setPosition(1)
      } else if (props.repeat === RP.none) {
        if (!props.lastTrack) {
          if (tick) {
            _queueNextTrack.current = true
          } else {
            props.nextTrack!()
            setPosition(0)
            setDuration(0)
          }
        } else {
          setPlaying(false)
        }
      }
    }
  }

  const onPlaying = () => {
    if (props.onPlaying) {
      props.onPlaying(position, duration)
    }
  }

  const onError = (component: string) => {
    console.error(component + ' - react-sound error')
  }

  const onPlay = () => {
    setPlaying(true)
  }

  const onPause = () => {
    setPlaying(false)
  }

  const onBack = () => {
    const amount = props.shorterSeek ? 5000 : 10000
    setPosition(Math.max(position - amount, 0))
  }

  const onForward = () => {
    const amount = props.shorterSeek ? 5000 : 10000
    setPosition(Math.min(position + amount, duration))
  }

  const { classes } = useStyles()
  const playStatus = playing
    ? (Sound as any).status.PLAYING
    : (Sound as any).status.PAUSED

  let msRemainder
  if (props.showMsTimestamp) {
    msRemainder = getMsRemainder(position)
    if (msRemainder === '.000') {
      msRemainder = undefined
    }
  }

  return (
    <React.Fragment key={props.audioID}>
      {props.audioEnabled && tick && playing && (
        <SoundTick
          url={url as string}
          playing={playStatus}
          speed={speed / 10}
          volume={volume}
          tick={tickState}
          onPlaying={onPlaying}
          onError={() => onError('SoundTick')}
          onFinishedPlaying={onFinishedPlaying}
        />
      )}
      {props.audioEnabled && !tick && (
        <Sound
          url={url as string}
          playStatus={playStatus}
          playbackRate={speed / 10}
          volume={volume}
          position={position}
          onPlaying={onPlaying}
          onError={() => onError('Sound')}
          onFinishedPlaying={onFinishedPlaying}
        />
      )}
      <Grid
        item
        xs={12}
        className={cx(!props.audioEnabled && classes.noPadding)}
      >
        <Collapse in={props.audioEnabled} className={classes.fullWidth}>
          <Grid container spacing={2} alignItems="center">
            <Grid item xs={12}>
              <Grid
                container
                spacing={1}
                alignItems="center"
                justifyContent="center"
              >
                <Grid item xs={12} sm={12}>
                  <Grid container spacing={1} alignItems="center">
                    <Grid item>
                      <Typography
                        variant="caption"
                        component="div"
                        color="textSecondary"
                      >
                        {getTimestampFromMs(position)}
                      </Typography>
                    </Grid>
                    <Grid item xs>
                      <Slider
                        valueLabelDisplay={msRemainder ? 'auto' : 'off'}
                        valueLabelFormat={msRemainder}
                        value={position}
                        classes={{
                          thumb: classes.noTransition,
                          track: classes.noTransition
                        }}
                        max={duration}
                        onChange={onChangePosition}
                      />
                    </Grid>
                    <Grid item>
                      <Typography
                        variant="caption"
                        component="div"
                        color="textSecondary"
                      >
                        {getTimestampFromMs(duration)}
                      </Typography>
                    </Grid>
                  </Grid>
                </Grid>
                <Grid item>
                  {props.prevTrack && (
                    <Tooltip disableInteractive title="Prev Track">
                      <IconButton onClick={props.prevTrack} size="large">
                        <SkipPreviousIcon />
                      </IconButton>
                    </Tooltip>
                  )}
                  <Tooltip disableInteractive title="Jump Back">
                    <IconButton onClick={onBack} size="large">
                      {props.shorterSeek ? <Replay5Icon /> : <Replay10Icon />}
                    </IconButton>
                  </Tooltip>
                  <Tooltip
                    disableInteractive
                    title={playing ? 'Pause' : 'Play'}
                  >
                    <IconButton
                      onClick={playing ? onPause : onPlay}
                      size="large"
                    >
                      {playing ? <PauseIcon /> : <PlayArrowIcon />}
                    </IconButton>
                  </Tooltip>
                  <Tooltip disableInteractive title="Jump Forward">
                    <IconButton onClick={onForward} size="large">
                      {props.shorterSeek ? <Forward5Icon /> : <Forward10Icon />}
                    </IconButton>
                  </Tooltip>
                  {props.nextTrack && (
                    <Tooltip disableInteractive title="Next Track">
                      <IconButton onClick={props.nextTrack} size="large">
                        <SkipNextIcon />
                      </IconButton>
                    </Tooltip>
                  )}
                </Grid>
              </Grid>
            </Grid>
            <Grid item xs={12}>
              <Grid container spacing={1} alignItems="center">
                <Grid item>
                  <VolumeDownIcon />
                </Grid>
                <Grid item xs>
                  <BaseSlider
                    selector={selectAudioVolume(props.audioID)}
                    action={props.audioVolumeAction}
                    labelledBy="audio-volume-slider"
                  />
                </Grid>
                <Grid item>
                  <VolumeUpIcon />
                </Grid>
              </Grid>
            </Grid>
          </Grid>
        </Collapse>
      </Grid>
    </React.Fragment>
  )
}

;(AudioControl as any).displayName = 'AudioControl'
export default AudioControl
