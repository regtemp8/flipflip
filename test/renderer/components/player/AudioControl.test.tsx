import React from "react";
import { describe, it, expect } from "@jest/globals";
import renderer from "react-test-renderer";
import AudioControl from "../../../../src/renderer/components/player/AudioControl";
import TestProvider from "../../../util/TestProvider";
import { RP } from "../../../../src/renderer/data/const";
import Audio from "../../../../src/renderer/data/Audio";

jest.mock('../../../../src/renderer/components/player/SoundTick', () => 'SoundTick');

// mocking this so that test doesn't throw error
jest.mock('@mui/material/Slider', () => 'Slider');

describe("AudioControl", () => {
  it("should match snapshot", () => {
    const audio = new Audio();
    const component = renderer.create(
      <TestProvider>
        <AudioControl
          audio={audio}
          audioEnabled={false}
          singleTrack={false}
          lastTrack={false}
          repeat={RP.none}
          scenePaths={[]}
          startPlaying={false}
          onAudioSliderChange={(e, value) => {}}
        />
      </TestProvider>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
