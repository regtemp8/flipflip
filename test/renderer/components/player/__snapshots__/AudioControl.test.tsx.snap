// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioControl should match snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 AudioControl-noPadding-2 css-13i4rnv-MuiGrid-root"
>
  <div
    className="MuiCollapse-root MuiCollapse-vertical AudioControl-fullWidth-1 MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
    style={
      {
        "minHeight": "0px",
      }
    }
  >
    <div
      className="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
    >
      <div
        className="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-ofrcno-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-z4xi2a-MuiGrid-root"
            >
              <div
                className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 css-1idn90j-MuiGrid-root"
              >
                <div
                  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-7ldjcq-MuiGrid-root"
                >
                  <div
                    className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
                  >
                    <div
                      className="MuiTypography-root MuiTypography-caption css-ol7wxs-MuiTypography-root"
                    >
                      0:00
                    </div>
                  </div>
                  <div
                    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                  >
                    <Slider
                      classes={
                        {
                          "thumb": "AudioControl-noTransition-3",
                          "track": "AudioControl-noTransition-3",
                        }
                      }
                      max={0}
                      onChange={[Function]}
                      value={0}
                      valueLabelDisplay="off"
                    />
                  </div>
                  <div
                    className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
                  >
                    <div
                      className="MuiTypography-root MuiTypography-caption css-ol7wxs-MuiTypography-root"
                    >
                      0:00
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <button
                  aria-label="Jump Back"
                  aria-labelledby={null}
                  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
                  data-mui-internal-clone-element={true}
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="Replay10Icon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11.99 5V1l-5 5 5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6h-2c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"
                    />
                    <path
                      d="M10.89 16h-.85v-3.26l-1.01.31v-.69l1.77-.63h.09V16zm4.28-1.76c0 .32-.03.6-.1.82s-.17.42-.29.57-.28.26-.45.33-.37.1-.59.1-.41-.03-.59-.1-.33-.18-.46-.33-.23-.34-.3-.57-.11-.5-.11-.82v-.74c0-.32.03-.6.1-.82s.17-.42.29-.57.28-.26.45-.33.37-.1.59-.1.41.03.59.1.33.18.46.33.23.34.3.57.11.5.11.82v.74zm-.85-.86c0-.19-.01-.35-.04-.48s-.07-.23-.12-.31-.11-.14-.19-.17-.16-.05-.25-.05-.18.02-.25.05-.14.09-.19.17-.09.18-.12.31-.04.29-.04.48v.97c0 .19.01.35.04.48s.07.24.12.32.11.14.19.17.16.05.25.05.18-.02.25-.05.14-.09.19-.17.09-.19.11-.32.04-.29.04-.48v-.97z"
                    />
                  </svg>
                  <span
                    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="Play"
                  aria-labelledby={null}
                  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
                  data-mui-internal-clone-element={true}
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="PlayArrowIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8 5v14l11-7z"
                    />
                  </svg>
                  <span
                    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="Jump Forward"
                  aria-labelledby={null}
                  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
                  data-mui-internal-clone-element={true}
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="Forward10Icon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M18 13c0 3.31-2.69 6-6 6s-6-2.69-6-6 2.69-6 6-6v4l5-5-5-5v4c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8h-2z"
                    />
                    <path
                      d="M10.86 15.94v-4.27h-.09L9 12.3v.69l1.01-.31v3.26zm1.39-2.5v.74c0 1.9 1.31 1.82 1.44 1.82.14 0 1.44.09 1.44-1.82v-.74c0-1.9-1.31-1.82-1.44-1.82-.14 0-1.44-.09-1.44 1.82zm2.04-.12v.97c0 .77-.21 1.03-.59 1.03s-.6-.26-.6-1.03v-.97c0-.75.22-1.01.59-1.01.38-.01.6.26.6 1.01z"
                    />
                  </svg>
                  <span
                    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-7ldjcq-MuiGrid-root"
            >
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="VolumeDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M18.5 12c0-1.77-1.02-3.29-2.5-4.03v8.05c1.48-.73 2.5-2.25 2.5-4.02zM5 9v6h4l5 5V4L9 9H5z"
                  />
                </svg>
              </div>
              <div
                className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
              >
                <Slider
                  aria-labelledby="audio-volume-slider"
                  onChange={[Function]}
                  value={100}
                />
              </div>
              <div
                className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="VolumeUpIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 9v6h4l5 5V4L7 9H3zm13.5 3c0-1.77-1.02-3.29-2.5-4.03v8.05c1.48-.73 2.5-2.25 2.5-4.02zM14 3.23v2.06c2.89.86 5 3.54 5 6.71s-2.11 5.85-5 6.71v2.06c4.01-.91 7-4.49 7-8.77s-2.99-7.86-7-8.77z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
